### PR TITLE
Implemented WorkflowRunStateChange schema

### PR DESCRIPTION
**Pull request by victorskl**
* Use opaque Payload object for differing workflow type
* Introduce two workflow types: `bssh_icav2_fastq_copy` and `bssh_bcl_convert`
* Introduce 3 namespaces for services: `orcabus.wfm`, `orcabus.bct`, `orcabus.fqcp`

### CODE REVIEW

**Comment by victorskl** [comment]
I still feel uneasy about Payload object being opaque about its attributes. It is practically making it un-controlled schema or domain object. 

Mechanism such as Schema Registry and Code binding (auto-generated code) missed this opaqueness. Hence, this would effectively dodge the schema control radar. If the upstream producer started to divergent from the initial payload structure, the subscriber may not aware of this. And, it'd ends up manual handle on event parsing.

But. Having said, let experiment this. This is what we discussed the other day.

**Comment by victorskl** [comment]
I have added [reading pointers](https://github.com/umccr/orcabus/issues/257#issuecomment-2095605777) to ticket. If you are able, highly recommended to give a read.

**Comment by victorskl** [comment]
Next, I will start prototyping baseline wfm #103 -- based on this WRSC. FYI

**Review by raylrui — APPROVED** [review]
LGTM!

**Comment by victorskl** [comment]
Self note on design: 

This is fine as we treat the all of Workflow business as one domain aggregate (DDD bounded context). Then, these are all internal events to this bounded context of the Workflow world and, events happening within it.

We can come up with another dedicated schema down the track; when externally observer would like to communicate into Workflow orchestration world events.

File: docs/event-schemas/wfm/WorkflowRunStateChange.json [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  "components": {
+    "schemas": {
+      "AWSEvent": {
+        "type": "object",
+        "required": [
+          "detail-type",
+          "detail",
+          "id",
+          "source",
+          "time"
```
Comment (reisingerf) [inline docs/event-schemas/wfm/WorkflowRunStateChange.json:17]: Should we make `account` mandatory? It's used in the archive filter.
Also, latest once we fix the payload schema(s), we should make the version mandatory.

File: docs/event-schemas/wfm/WorkflowRunStateChange.json [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      "WorkflowRunStateChange": {
+        "type": "object",
+        "required": [
+          "portalRunId",
+          "timestamp",
+          "status",
+          "workflowType",
+          "workflowVersion",
+          "payload",
+          "serviceVersion"
```
Comment (reisingerf) [inline docs/event-schemas/wfm/WorkflowRunStateChange.json:62]: Do we need a `service` or is the `source` sufficient?

**Review by reisingerf — APPROVED** [review]

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 2 above)
Comment (alexiswl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:62]: What's the difference between service and source?

**Comment by alexiswl** [comment]
> I still feel uneasy about Payload object being opaque about its attributes. It is practically making it un-controlled schema or domain object.

I agree but I don't think this is 'forever', good to get started with something and we can move to something more robust later. 

I think once we're up and running with our set services we can move this to something like

```
{
  "payload": {
    "type": "object",
   "properties": "id",
   "oneOf": [
      {
        "$ref": "#definitions/microserviceA"
      },
      {
        "$ref": "#definitions/microserviceB"
      }
   ]
  }
}
```

**Comment by victorskl** [comment]
Yup. Thought about that too, Alexis.

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 2 above)
Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:62]: I think Flo is referring to `serviceVersion`

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 1 above)
Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:17]: Yup. We can do that.

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 2 above)
Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:62]: > The service differentiator marker on events is the `source` property of the event envelope.

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 1 above)
Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:17]: Final AWS Envelope:

We did run experiment and, turned out `id` and `version` are not in our control. EventBus fills these value regardless. Also, `version` seems to be always set to `0` on the actual event instance that sent into the channel; though we payload it with `1.0.0`. 

The `account`, `region`, `time` observed to be auto-filled by EventBridge as well.

Hence. We concluded that we will be just doing the required fields on `detail`, `detail-type`, `source` as per AWS documentation. See below.

Minimum information needed for a valid custom event
https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-events-structure.html

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 2 above)
Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:62]: We discussed about pros/cons of opaqueness of Payload object. And, tried tinker with alternate approaches such as fixing the schemas upfront.

For now, we concluded we shall keep it as Payload object.

The `serviceVersion` will be refactored to `version` property of the `Payload` definition as the required field; along with payload `refId`.

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 1 above)
Comment (raylrui) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:17]: And some more reading on this:
https://docs.aws.amazon.com/AmazonS3/latest/userguide/ev-events.html
The version seems applied in "Create a schema from events on an event bus"
>If the contents of events on the event bus change, EventBridge creates new versions of the related EventBridge schema.
https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-schema-create.html

Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json]: Updated with https://github.com/umccr/orcabus/commit/1e903c67c1eb0fe8ffa949171a058c93c835c4d7

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 2 above)
Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json:60]: Updated with https://github.com/umccr/orcabus/commit/1e903c67c1eb0fe8ffa949171a058c93c835c4d7

**Comment by victorskl** [comment]
@alexiswl @williamputraintan tick pls if you are ok with PR

**Review by alexiswl — APPROVED** [review]

**Review by williamputraintan — APPROVED** [review]

File: docs/event-schemas/wfm/WorkflowRunStateChange.json (hunk 1 above)
Comment (victorskl) [reply docs/event-schemas/wfm/WorkflowRunStateChange.json]: Also harmonised/updated SRSC envelope with https://github.com/umccr/orcabus/commit/34904419b98e20498a02a65c685c93f2ba93993d

**Comment by victorskl** [comment]
Alright. Good luck with our very first schema.! 🤞 This is the start...

**Pull request merged**